Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,226 @@
+        current = dynamicDowncast<SVGPatternElement>(target.element.get());
+    }
+
+    m_attributes = WTFMove(attributes);
+}
+
+RefPtr<Pattern> RenderSVGResourcePattern::buildPattern(GraphicsContext& context, const RenderLayerModelObject& renderer)
+{
+    // Spec: When the geometry of the applicable element has no width or height and objectBoundingBox is specified,
+    // then the given effect (e.g. a gradient or a filter) will be ignored.
+    FloatRect objectBoundingBox = renderer.objectBoundingBox();
+    if (m_attributes->patternUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX && objectBoundingBox.isEmpty())
+        return nullptr;
+
+    // If we couldn't determine the pattern content element root, stop here.
+    if (!m_attributes->patternContentElement())
+        return nullptr;
+
+    // An empty viewBox disables rendering.
+    if (m_attributes->hasViewBox() && m_attributes->viewBox().isEmpty())
+        return nullptr;
+
+    // Compute all necessary transformations to build the tile image & the pattern.
+    FloatRect tileBoundaries;
+    AffineTransform tileImageTransform;
+    if (!buildTileImageTransform(renderer, *m_attributes, patternElement(), tileBoundaries, tileImageTransform))
+        return nullptr;
+
+    // Ignore 2D rotation, as it doesn't affect the size of the tile.
+    auto absoluteTransformIgnoringRotation = context.getCTM(GraphicsContext::DefinitelyIncludeDeviceScale);
+    SVGRenderingContext::clear2DRotation(absoluteTransformIgnoringRotation);
+    FloatRect absoluteTileBoundaries = absoluteTransformIgnoringRotation.mapRect(tileBoundaries);
+
+    // Scale the tile size to match the scale level of the patternTransform.
+    absoluteTileBoundaries.scale(static_cast<float>(m_attributes->patternTransform().xScale()), static_cast<float>(m_attributes->patternTransform().yScale()));
+
+    // Build tile image.
+    auto tileImage = createTileImage(*m_attributes, tileBoundaries, absoluteTileBoundaries, tileImageTransform);
+    if (!tileImage)
+        return nullptr;
+
+    auto tileImageSize = tileImage->logicalSize();
+
+    auto copiedImage = ImageBuffer::sinkIntoNativeImage(WTFMove(tileImage));
+    if (!copiedImage)
+        return nullptr;
+
+    // Compute pattern space transformation.
+    AffineTransform patternSpaceTransform;
+    patternSpaceTransform.translate(tileBoundaries.location());
+    patternSpaceTransform.scale(tileBoundaries.size() / tileImageSize);
+
+    AffineTransform patternTransform = m_attributes->patternTransform();
+    if (!patternTransform.isIdentity())
+        patternSpaceTransform = patternTransform * patternSpaceTransform;
+
+    // Build pattern.
+    return Pattern::create({ copiedImage.releaseNonNull() }, { true, true, patternSpaceTransform });
+}
+
+bool RenderSVGResourcePattern::buildPatternIfNeeded(GraphicsContext& context, const RenderLayerModelObject& targetRenderer)
+{
+    collectPatternAttributesIfNeeded();
```

Comment:
I think for gradients this returns a bool, false indicating early exit due to empty bbox- why not mirror it 1:1 here

Reply:
I made collectPatternAttributesIfNeeded do more checks.